File: src/com/mojang/ld22/save/EntityIO.java
package com.mojang.ld22.save;

import com.google.gwt.json.client.JSONArray;
import com.google.gwt.json.client.JSONNull;
import com.google.gwt.json.client.JSONNumber;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONString;
import com.google.gwt.json.client.JSONValue;
import com.mojang.ld22.Game;
import com.mojang.ld22.entity.AirWizard;
import com.mojang.ld22.entity.Anvil;
import com.mojang.ld22.entity.Chest;
import com.mojang.ld22.entity.Entity;
import com.mojang.ld22.entity.Furniture;
import com.mojang.ld22.entity.Inventory;
import com.mojang.ld22.entity.ItemEntity;
import com.mojang.ld22.entity.Lantern;
import com.mojang.ld22.entity.Mob;
import com.mojang.ld22.entity.Oven;
import com.mojang.ld22.entity.Player;
import com.mojang.ld22.entity.Slime;
import com.mojang.ld22.entity.Workbench;
import com.mojang.ld22.entity.Zombie;
import com.mojang.ld22.item.Item;

/**
 * Multi-type serializer/deserializer for the entities that are persisted.
 *
 * <p>typeId assignment: 1=Player, 2=Zombie, 3=Slime, 4=AirWizard,
 * 6=Chest, 7=Workbench, 8=Anvil, 9=Furnace, 10=Oven, 11=Lantern, 12=ItemEntity.
 * (5 and 13+ reserved; {@code Spark} is intentionally never written.)
 *
 * <p>When a new serializable entity is added, it MUST be registered here with a
 * stable typeId and covered by a round-trip test.
 */
public final class EntityIO {

    private EntityIO() {
    }

    // ------------------------------------------------------------------
    // WRITE
    // ------------------------------------------------------------------

    public static JSONObject write(Entity e) {
        JSONObject o = new JSONObject();
        if (e instanceof Player) {
            o.put("typeId", new JSONNumber(1));
            writeMob((Mob) e, o);
            writePlayer((Player) e, o);
        } else if (e instanceof Zombie) {
            o.put("typeId", new JSONNumber(2));
            writeMob((Mob) e, o);
            o.put("lvl", new JSONNumber(((Zombie) e).lvl));
        } else if (e instanceof Slime) {
            o.put("typeId", new JSONNumber(3));
            writeMob((Mob) e, o);
            o.put("lvl", new JSONNumber(((Slime) e).lvl));
        } else if (e instanceof AirWizard) {
            o.put("typeId", new JSONNumber(4));
            writeMob((Mob) e, o);
        } else if (e instanceof Chest) {
            o.put("typeId", new JSONNumber(6));
            writeFurniture(e, o);
            writeInventory(((Chest) e).inventory, o);
        } else if (e instanceof Workbench) {
            o.put("typeId", new JSONNumber(7));
            writeFurniture(e, o);
        } else if (e instanceof Anvil) {
            o.put("typeId", new JSONNumber(8));
            writeFurniture(e, o);
        } else if (e instanceof Furnace) {
            o.put("typeId", new JSONNumber(9));
            writeFurniture(e, o);
        } else if (e instanceof Oven) {
            o.put("typeId", new JSONNumber(10));
            writeFurniture(e, o);
        } else if (e instanceof Lantern) {
            o.put("typeId", new JSONNumber(11));
            writeFurniture(e, o);
        } else if (e instanceof ItemEntity) {
            o.put("typeId", new JSONNumber(12));
            writeEntityBase(e, o);
            o.put("item", ItemIO.write(((ItemEntity) e).item));
            o.put("lifeTime", new JSONNumber(((ItemEntity) e).lifeTime));
            o.put("time", new JSONNumber(((ItemEntity) e).time));
        }
        return o;
    }

    private static void writeEntityBase(Entity e, JSONObject o) {
        o.put("x", new JSONNumber(e.x));
        o.put("y", new JSONNumber(e.y));
        o.put("xr", new JSONNumber(e.xr));
        o.put("yr", new JSONNumber(e.yr));
    }

    private static void writeMob(Mob m, JSONObject o) {
        writeEntityBase(m, o);
        o.put("walkDist", new JSONNumber(m.walkDist));
        o.put("dir", new JSONNumber(m.dir));
        o.put("hurtTime", new JSONNumber(m.hurtTime));
        o.put("xKnockback", new JSONNumber(m.xKnockback));
        o.put("yKnockback", new JSONNumber(m.yKnockback));
        o.put("maxHealth", new JSONNumber(m.maxHealth));
        o.put("health", new JSONNumber(m.health));
        o.put("swimTimer", new JSONNumber(m.swimTimer));
        o.put("tickTime", new JSONNumber(m.tickTime));
    }

    private static void writeFurniture(Entity e, JSONObject o) {
        writeEntityBase(e, o);
        Furniture f = (Furniture) e;
        o.put("name", new JSONString(f.name));
        o.put("col", new JSONNumber(f.col));
        o.put("sprite", new JSONNumber(f.sprite));
        o.put("pushTime", new JSONNumber(f.pushTime));
        o.put("pushDir", new JSONNumber(f.pushDir));
    }

    private static void writePlayer(Player p, JSONObject o) {
        writeInventory(p.inventory, o);
        o.put("attackItem", p.attackItem == null ? JSONNull.getInstance() : ItemIO.write(p.attackItem));
        o.put("activeItem", p.activeItem == null ? JSONNull.getInstance() : ItemIO.write(p.activeItem));
        o.put("stamina", new JSONNumber(p.stamina));
        o.put("staminaRecharge", new JSONNumber(p.staminaRecharge));
        o.put("staminaRechargeDelay", new JSONNumber(p.staminaRechargeDelay));
        o.put("score", new JSONNumber(p.score));
        o.put("maxStamina", new JSONNumber(p.maxStamina));
        o.put("invulnerableTime", new JSONNumber(p.invulnerableTime));
    }

    private static void writeInventory(Inventory inv, JSONObject o) {
        JSONArray arr = new JSONArray();
        for (int i = 0; i < inv.items.size(); i++) {
            arr.set(i, ItemIO.write(inv.items.get(i)));
        }
        o.put("inventory", arr);
    }

    // ------------------------------------------------------------------
    // READ
    // ------------------------------------------------------------------

    public static Entity read(JSONObject o, Game game) {
        if (o == null) return null;
        int typeId = (int) o.get("typeId").isNumber().doubleValue();
        switch (typeId) {
            case 1: {
                Player p = new Player(game, game.input, false); // do NOT re-seed starting items
                readMob(p, o);
                readPlayer(p, o);
                return p;
            }
            case 2: {
                Zombie z = new Zombie(1);
                readMob(z, o);
                z.lvl = (int) o.get("lvl").isNumber().doubleValue();
                return z;
            }
            case 3: {
                Slime s = new Slime(1);
                readMob(s, o);
                s.lvl = (int) o.get("lvl").isNumber().doubleValue();
                return s;
            }
            case 4: {
                AirWizard a = new AirWizard();
                readMob(a, o);
                return a;
            }
            case 6: {
                Chest c = new Chest();
                readFurniture(c, o);
                readInventory(c.inventory, o);
                return c;
            }
            case 7:
                return new Workbench();
            case 8:
                return new Anvil();
            case 9:
                return new Furnace();
            case 10:
                return new Oven();
            case 11:
                return new Lantern();
            case 12: {
                Item it = ItemIO.read(o.get("item").isObject());
                if (it == null) return null; // unresolvable drop item: skip this entity
                ItemEntity ie = new ItemEntity(it, 0, 0);
                readEntityBase(ie, o);
                ie.lifeTime = (int) o.get("lifeTime").isNumber().doubleValue();
                ie.time = (int) o.get("time").isNumber().doubleValue();
                return ie;
            }
            default:
                return null;
        }
    }

    private static void readEntityBase(Entity e, JSONObject o) {
        e.x = (int) o.get("x").isNumber().doubleValue();
        e.y = (int) o.get("y").isNumber().doubleValue();
        e.xr = (int) o.get("xr").isNumber().doubleValue();
        e.yr = (int) o.get("yr").isNumber().doubleValue();
    }

    private static void readMob(Mob m, JSONObject o) {
        readEntityBase(m, o);
        m.walkDist = (int) o.get("walkDist").isNumber().doubleValue();
        m.dir = (int) o.get("dir").isNumber().doubleValue();
        m.hurtTime = (int) o.get("hurtTime").isNumber().doubleValue();
        m.xKnockback = (int) o.get("xKnockback").isNumber().doubleValue();
        m.yKnockback = (int) o.get("yKnockback").isNumber().doubleValue();
        m.maxHealth = (int) o.get("maxHealth").isNumber().doubleValue();
        m.health = (int) o.get("health").isNumber().doubleValue();
        m.swimTimer = (int) o.get("swimTimer").isNumber().doubleValue();
        m.tickTime = (int) o.get("tickTime").isNumber().doubleValue();
    }

    private static void readFurniture(Furniture f, JSONObject o) {
        readEntityBase(f, o);
        f.name = o.get("name").isString().stringValue();
        f.col = (int) o.get("col").isNumber().doubleValue();
        f.sprite = (int) o.get("sprite").isNumber().doubleValue();
        f.pushTime = (int) o.get("pushTime").isNumber().doubleValue();
        f.pushDir = (int) o.get("pushDir").isNumber().doubleValue();
    }

    private static void readPlayer(Player p, JSONObject o) {
        readInventory(p.inventory, o);
        JSONValue atk = o.get("attackItem");
        p.attackItem = (atk == null || atk.isNull() != null) ? null : ItemIO.read(atk.isObject());
        JSONValue act = o.get("activeItem");
        p.activeItem = (act == null || act.isNull() != null) ? null : ItemIO.read(act.isObject());
        p.stamina = (int) o.get("stamina").isNumber().doubleValue();
        p.staminaRecharge = (int) o.get("staminaRecharge").isNumber().doubleValue();
        p.staminaRechargeDelay = (int) o.get("staminaRechargeDelay").isNumber().doubleValue();
        p.score = (int) o.get("score").isNumber().doubleValue();
        p.maxStamina = (int) o.get("maxStamina").isNumber().doubleValue();
        p.invulnerableTime = (int) o.get("invulnerableTime").isNumber().doubleValue();
    }

    private static void readInventory(Inventory inv, JSONObject o) {
        JSONArray arr = o.get("inventory").isArray();
        if (arr == null) return;
        for (int i = 0; i < arr.size(); i++) {
            JSONValue v = arr.get(i);
            if (v == null || v.isNull() != null) continue;
            Item it = ItemIO.read(v.isObject());
            if (it != null) inv.items.add(it); // add directly: preserve order & stacks
        }
    }
}

File: ts-port/src/game/save/EntityIO.ts
/**
 * EntityIO — serialises the persisted entity set to/from plain JSON objects.
 *
 * Dispatch key is `t` (typeId):
 *   1  Player
 *   2  Zombie      (stores `lvl`)
 *   3  Slime       (stores `lvl`)
 *   4  AirWizard
 *   6  Chest       (stores inventory)
 *   7  Workbench
 *   8  Anvil
 *   9  Furnace
 *   10 Oven
 *   11 Lantern
 *   12 ItemEntity  (dropped item; stores the item + remaining lifetime)
 *
 * 5 (Spark) and all particles are transient and are never written — `write`
 * returns null for anything without a persistent representation, and the caller
 * skips nulls. Mob common fields (position, health, knockback, timers) are
 * captured by `writeMobCommon`/`applyMobCommon`; per-type extras are layered on
 * top. Attack timers (AirWizard) and the player's private attackTime/attackDir
 * are intentionally omitted — they reset to a clean state on load.
 */
import type { Entity } from '../entity/Entity';
import type { Game } from '../Game';
import type { Mob } from '../entity/Mob';
import { Player } from '../entity/Player';
import { Zombie } from '../entity/Zombie';
import { Slime } from '../entity/Slime';
import { AirWizard } from '../entity/AirWizard';
import { Chest } from '../entity/Chest';
import { Furniture } from '../entity/Furniture';
import { Workbench } from '../entity/Workbench';
import { Anvil } from '../entity/Anvil';
import { Furnace } from '../entity/Furnace';
import { Oven } from '../entity/Oven';
import { Lantern } from '../entity/Lantern';
import { ItemEntity } from '../entity/ItemEntity';
import type { ResourceItem } from '../item/ResourceItem';
import { ItemIO } from './ItemIO';

interface MobPayload {
  x: number;
  y: number;
  xr: number; // P2-8: persisted radial hitbox (Java EntityIO.writeEntityBase)
  yr: number;
  walkDist: number;
  dir: number;
  hurtTime: number;
  xKnockback: number;
  yKnockback: number;
  maxHealth: number;
  health: number;
  swimTimer: number;
  tickTime: number;
}

function writeMobCommon(m: Mob): MobPayload {
  return {
    x: m.x,
    y: m.y,
    xr: m.xr,
    yr: m.yr,
    walkDist: m.walkDist,
    dir: m.dir,
    hurtTime: m.hurtTime,
    xKnockback: m.xKnockback,
    yKnockback: m.yKnockback,
    maxHealth: m.maxHealth,
    health: m.health,
    swimTimer: m.swimTimer,
    tickTime: m.tickTime,
  };
}

function applyMobCommon(m: Mob, o: MobPayload): void {
  m.x = o.x;
  m.y = o.y;
  m.xr = o.xr ?? 4; // P2-8: matches Mob ctor default (xr=4, yr=3)
  m.yr = o.yr ?? 3;
  m.walkDist = o.walkDist ?? 0;
  m.dir = o.dir ?? 0;
  m.hurtTime = o.hurtTime ?? 0;
  m.xKnockback = o.xKnockback ?? 0;
  m.yKnockback = o.yKnockback ?? 0;
  m.maxHealth = o.maxHealth ?? 10;
  m.health = o.health ?? m.maxHealth;
  m.swimTimer = o.swimTimer ?? 0;
  m.tickTime = o.tickTime ?? 0;
}

function writeInventory(inv: { items: unknown[] }): unknown[] {
  const out: unknown[] = [];
  for (const it of inv.items) {
    const w = ItemIO.write(it as never);
    if (w) out.push(w);
  }
  return out;
}

function readInventoryInto(target: { items: unknown[] }, arr: unknown): void {
  target.items.length = 0; // start clean; direct push preserves exact slots
  if (!Array.isArray(arr)) return;
  for (const o of arr) {
    const item = ItemIO.read(o);
    if (item) target.items.push(item);
  }
}

export class EntityIO {
  /** Serialise an entity, or null if it has no persistent representation. */
  public static write(e: Entity): Record<string, unknown> | null {
    if (e instanceof Player) {
      const base = writeMobCommon(e);
      return {
        t: 1,
        ...base,
        score: e.score,
        maxStamina: e.maxStamina,
        stamina: e.stamina,
        invulnerableTime: e.invulnerableTime,
        inventory: writeInventory(e.inventory),
        activeItem: e.activeItem ? ItemIO.write(e.activeItem) : null,
        attackItem: e.attackItem ? ItemIO.write(e.attackItem) : null,
      };
    }
    if (e instanceof Zombie) {
      return { t: 2, ...writeMobCommon(e), lvl: e.lvl };
    }
    if (e instanceof Slime) {
      return { t: 3, ...writeMobCommon(e), lvl: e.lvl };
    }
    if (e instanceof AirWizard) {
      return { t: 4, ...writeMobCommon(e) };
    }
    if (e instanceof Chest) {
      return {
        t: 6,
        x: e.x,
        y: e.y,
        col: e.col,
        sprite: e.sprite,
        name: e.name,
        pushTime: e.pushTime, // P2-7
        pushDir: e.pushDir, // P2-7
        inventory: writeInventory(e.inventory),
      };
    }
    if (e instanceof Workbench) {
      return { t: 7, x: e.x, y: e.y, col: e.col, sprite: e.sprite, name: e.name, pushTime: e.pushTime, pushDir: e.pushDir };
    }
    if (e instanceof Anvil) {
      return { t: 8, x: e.x, y: e.y, col: e.col, sprite: e.sprite, name: e.name, pushTime: e.pushTime, pushDir: e.pushDir };
    }
    if (e instanceof Furnace) {
      return { t: 9, x: e.x, y: e.y, col: e.col, sprite: e.sprite, name: e.name, pushTime: e.pushTime, pushDir: e.pushDir };
    }
    if (e instanceof Oven) {
      return { t: 10, x: e.x, y: e.y, col: e.col, sprite: e.sprite, name: e.name, pushTime: e.pushTime, pushDir: e.pushDir };
    }
    if (e instanceof Lantern) {
      return { t: 11, x: e.x, y: e.y, col: e.col, sprite: e.sprite, name: e.name, pushTime: e.pushTime, pushDir: e.pushDir };
    }
    if (e instanceof ItemEntity) {
      // The dropped item is always a ResourceItem in the slice; persist it so
      // ground loot survives a save/load instead of vanishing.
      return {
        t: 12,
        x: e.x,
        y: e.y,
        lifeTime: e.lifeTime,
        time: e.time,
        hurtTime: e.hurtTime,
        item: ItemIO.write(e.item),
      };
    }
    // Spark, particles and any unknown entity: do not persist.
    return null;
  }

  /** Reconstruct an entity from its payload, or null if unrecognised. */
  public static read(o: unknown, game: Game): Entity | null {
    if (!o || typeof o !== 'object') return null;
    const obj = o as Record<string, unknown> & MobPayload;
    switch (obj.t) {
      case 1: {
        const p = new Player(game, game.input);
        applyMobCommon(p, obj);
        p.score = (obj.score as number) ?? 0;
        p.maxStamina = (obj.maxStamina as number) ?? 10;
        p.stamina = (obj.stamina as number) ?? p.maxStamina;
        p.invulnerableTime = (obj.invulnerableTime as number) ?? 0;
        readInventoryInto(p.inventory, obj.inventory);
        p.activeItem = obj.activeItem ? ItemIO.read(obj.activeItem) : null;
        p.attackItem = obj.attackItem ? ItemIO.read(obj.attackItem) : null;
        return p;
      }
      case 2: {
        const z = new Zombie((obj.lvl as number) ?? 1);
        applyMobCommon(z, obj);
        return z;
      }
      case 3: {
        const s = new Slime((obj.lvl as number) ?? 1);
        applyMobCommon(s, obj);
        return s;
      }
      case 4: {
        const aw = new AirWizard();
        applyMobCommon(aw, obj);
        return aw;
      }
      case 6: {
        const c = new Chest();
        c.x = obj.x as number;
        c.y = obj.y as number;
        c.col = obj.col as number;
        c.sprite = obj.sprite as number;
        c.name = obj.name as string;
        c.pushTime = (obj.pushTime as number) ?? 0; // P2-7
        c.pushDir = (obj.pushDir as number) ?? -1; // P2-7
        readInventoryInto(c.inventory, obj.inventory);
        return c;
      }
      case 7:
      case 8:
      case 9:
      case 10:
      case 11: {
        const f = Furniture.createByName(obj.name as string);
        if (!f) return null;
        f.x = obj.x as number;
        f.y = obj.y as number;
        f.col = obj.col as number;
        f.sprite = obj.sprite as number;
        f.name = obj.name as string;
        f.pushTime = (obj.pushTime as number) ?? 0; // P2-7
        f.pushDir = (obj.pushDir as number) ?? -1; // P2-7
        return f;
      }
      case 12: {
        const item = ItemIO.read(obj.item) as ResourceItem | null;
        if (!item) return null;
        const ie = new ItemEntity(item, obj.x as number, obj.y as number);
        ie.lifeTime = (obj.lifeTime as number) ?? ie.lifeTime;
        ie.time = (obj.time as number) ?? ie.time;
        ie.hurtTime = (obj.hurtTime as number) ?? 0;
        return ie;
      }
      default:
        return null;
    }
  }
}
